# 10. Applied Data Science Capstone

### Peer-graded Assignment - Week 3

# Libraries

In [1]:
# import libraries

# for data handling and analysis
import numpy as np
import pandas as pd

# for webscraping
from bs4 import BeautifulSoup

# to convert address to long-lat
!pip install geopy
from geopy.geocoders import Nominatim

# to handle requests with Foursquare API
import requests

# for plotting
import matplotlib.cm as cm
import matplotlib.colors as colors 

# for k-means clustering
from sklearn.cluster import KMeans

# for rendering maps
!conda install -c conda-forge folium=0.5.0 --yes
import folium

print('Libraries are good to go.')


Solving environment: ...working... done

# All requested packages already installed.

Libraries are good to go.


# Part I - Building the Dataframe

In [2]:
# url where the table with data is
url = 'https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M'
html_data = requests.get(url).text

In [3]:
# create a beautiful soup object
soup = BeautifulSoup(html_data, 'html5lib')

# check for readable text
#print(soup.prettify())

In [4]:
# see how many tables on the url
tables = soup.find_all('table')                     
len(tables)

3

In [5]:
# find the index of table that has the relevant data

for index,table in enumerate(tables):                 
    if ("List_of_postal_codes_of_Canada" in str(table)):
        table_index = index
print(table_index)

2


In [6]:
# extract table 

# create an empty list
table_contents=[]

# find the table and write to the empty list, ingoring 'Not assigned' boroughs
# there will be one line for each post code, with multiple neighbourhoods split by a comma
table=soup.find('table')
for row in table.findAll('td'):
    cell = {}
    if row.span.text=='Not assigned':
        pass
    else:
        cell['PostalCode'] = row.p.text[:3]
        cell['Borough'] = (row.span.text).split('(')[0]
        cell['Neighborhood'] = (((((row.span.text).split('(')[1]).strip(')')).replace(' /',',')).replace(')',' ')).strip(' ')
        table_contents.append(cell)


In [7]:
# convert the table_contents list to a dataframe
df=pd.DataFrame(table_contents)

### Some of the borough names are odd and replaced with something that makes more sense.

In [8]:
df['Borough']=df['Borough'].replace({'Downtown TorontoStn A PO Boxes25 The Esplanade':'Downtown Toronto Stn A',
                                             'East TorontoBusiness reply mail Processing Centre969 Eastern':'East Toronto Business',
                                             'EtobicokeNorthwest':'Etobicoke Northwest','East YorkEast Toronto':'East York/East Toronto',
                                             'MississaugaCanada Post Gateway Processing Centre':'Mississauga'})

df.head(12)

PostalCode           Borough  \
0         M3A        North York   
1         M4A        North York   
2         M5A  Downtown Toronto   
3         M6A        North York   
4         M7A      Queen's Park   
5         M9A         Etobicoke   
6         M1B       Scarborough   
7         M3B        North York   
8         M4B         East York   
9         M5B  Downtown Toronto   
10        M6B        North York   
11        M9B         Etobicoke   

                                         Neighborhood  
0                                           Parkwoods  
1                                    Victoria Village  
2                           Regent Park, Harbourfront  
3                    Lawrence Manor, Lawrence Heights  
4                       Ontario Provincial Government  
5                                    Islington Avenue  
6                                      Malvern, Rouge  
7                                     Don Mills North  
8                     Parkview Hill, Woodbine Gardens  
9                            Garden District, Ryerson  
10                                          Glencairn  
11  West Deane Park, Princess Gardens, Martin Grov...

### Finally check the shape of the dataframe

In [9]:
print('Rows in the dataframe: ', df.shape[0]) 
print('Columns in the dataframe: ', df.shape[1])

Rows in the dataframe:  103
Columns in the dataframe:  3


# Part II - Adding Coordinates

### The geocoder package did not work for me. Foursquare nearly did. But in the end I went with the csv file provided.

In [10]:
df_coords = pd.read_csv('C:/Users/henny/OneDrive/Courses/IBM Data Science/10_applied_data_science_capstone/Geospatial_Coordinates.csv')
df_coords.head()

Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

### Renaming the postal code column so it is identical in both dataframes for easy merging.

In [11]:
df_coords.rename(columns={"Postal Code" : "PostalCode"}, inplace=True)
df_coords.head()

PostalCode   Latitude  Longitude
0        M1B  43.806686 -79.194353
1        M1C  43.784535 -79.160497
2        M1E  43.763573 -79.188711
3        M1G  43.770992 -79.216917
4        M1H  43.773136 -79.239476

### Merging latitude and longitude data to the original dataframe.

In [12]:
df_with_coords = pd.merge(df, df_coords, how="left")
df_with_coords.head(12)

PostalCode           Borough  \
0         M3A        North York   
1         M4A        North York   
2         M5A  Downtown Toronto   
3         M6A        North York   
4         M7A      Queen's Park   
5         M9A         Etobicoke   
6         M1B       Scarborough   
7         M3B        North York   
8         M4B         East York   
9         M5B  Downtown Toronto   
10        M6B        North York   
11        M9B         Etobicoke   

                                         Neighborhood   Latitude  Longitude  
0                                           Parkwoods  43.753259 -79.329656  
1                                    Victoria Village  43.725882 -79.315572  
2                           Regent Park, Harbourfront  43.654260 -79.360636  
3                    Lawrence Manor, Lawrence Heights  43.718518 -79.464763  
4                       Ontario Provincial Government  43.662301 -79.389494  
5                                    Islington Avenue  43.667856 -79.532242  
6                                      Malvern, Rouge  43.806686 -79.194353  
7                                     Don Mills North  43.745906 -79.352188  
8                     Parkview Hill, Woodbine Gardens  43.706397 -79.309937  
9                            Garden District, Ryerson  43.657162 -79.378937  
10                                          Glencairn  43.709577 -79.445073  
11  West Deane Park, Princess Gardens, Martin Grov...  43.650943 -79.554724

# Part III - Explore and Cluster the Neighbourhoods

### Let's explore the data a little bit. First we look at how many boroughs and neighbourhoods.

In [13]:
print('The dataframe has {} boroughs and {} neighbourhoods'.format(
        len(df_with_coords['Borough'].unique()),
        df_with_coords.shape[0]
    )
)

The dataframe has 15 boroughs and 103 neighbourhoods


### Check the datatypes.

In [14]:
df_with_coords.dtypes

PostalCode       object
Borough          object
Neighborhood     object
Latitude        float64
Longitude       float64
dtype: object

### Check for missing data.

In [15]:
df_with_coords.isnull().sum()

PostalCode      0
Borough         0
Neighborhood    0
Latitude        0
Longitude       0
dtype: int64

### Have a look at the ranges.

In [16]:
df_with_coords.describe()

Latitude   Longitude
count  103.000000  103.000000
mean    43.704608  -79.397153
std      0.052463    0.097146
min     43.602414  -79.615819
25%     43.660567  -79.464763
50%     43.696948  -79.388790
75%     43.745320  -79.340923
max     43.836125  -79.160497

### Create a map to have some intuition. Use geopy for latlong values and folium for the map. 

In [17]:
# Use geopy library to get the latitude and longitude values of Toronto.
address = 'Toronto, Canada'

geolocator = Nominatim(user_agent="toronto_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinates of Toronto are {}, {}.'.format(latitude, longitude))

The geograpical coordinates of Toronto are 43.6534817, -79.3839347.


In [18]:
# create map of Toronto using latitude and longitude values
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, borough, neighborhood in zip(df_with_coords['Latitude'], df_with_coords['Longitude'], df_with_coords['Borough'], df_with_coords['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

In [19]:
# @hidden_cell
CLIENT_ID = 'DECS0BLENBNCUVUZFOEQCLIPZRXOWDVHEWV2I4ACPKHMKEC4' 
CLIENT_SECRET = 'L1LUXMTOZZEQQX35YPEIO4GTB0VIWFX2VW14AHFLCZBIUI1M' 
VERSION = '20180605' 

### And explore the top 100 venues within a 500 meters radius.

In [20]:
LIMIT = 100
RADIUS = 500

### create code for all neighbourhoods

In [21]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            RADIUS, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [22]:
# create a dataframe for Toronto venues
toronto_venues = getNearbyVenues(names=df_with_coords['Neighborhood'],
                                latitudes=df_with_coords['Latitude'],
                                longitudes=df_with_coords['Longitude']
                                )


Parkwoods
Victoria Village
Regent Park, Harbourfront
Lawrence Manor, Lawrence Heights
Ontario Provincial Government
Islington Avenue
Malvern, Rouge
Don Mills North
Parkview Hill, Woodbine Gardens
Garden District, Ryerson
Glencairn
West Deane Park, Princess Gardens, Martin Grove, Islington, Cloverdale
Rouge Hill, Port Union, Highland Creek
Don Mills South
Woodbine Heights
St. James Town
Humewood-Cedarvale
Eringate, Bloordale Gardens, Old Burnhamthorpe, Markland Wood
Guildwood, Morningside, West Hill
The Beaches
Berczy Park
Caledonia-Fairbanks
Woburn
Leaside
Central Bay Street
Christie
Cedarbrae
Hillcrest Village
Bathurst Manor, Wilson Heights, Downsview North
Thorncliffe Park
Richmond, Adelaide, King
Dufferin, Dovercourt Village
Scarborough Village
Fairview, Henry Farm, Oriole
Northwood Park, York University
The Danforth  East
Harbourfront East, Union Station, Toronto Islands
Little Portugal, Trinity
Kennedy Park, Ionview, East Birchmount Park
Bayview Village
Downsview East
The Danforth

### Let's explore the dataframe

In [23]:
print(toronto_venues.shape)
toronto_venues.head()

(2141, 7)


Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0         Parkwoods              43.753259              -79.329656   
1         Parkwoods              43.753259              -79.329656   
2         Parkwoods              43.753259              -79.329656   
3  Victoria Village              43.725882              -79.315572   
4  Victoria Village              43.725882              -79.315572   

                    Venue  Venue Latitude  Venue Longitude  \
0         Brookbanks Park       43.751976       -79.332140   
1                     KFC       43.754387       -79.333021   
2           Variety Store       43.751974       -79.333114   
3  Victoria Village Arena       43.723481       -79.315635   
4             Tim Hortons       43.725517       -79.313103   

         Venue Category  
0                  Park  
1  Fast Food Restaurant  
2     Food & Drink Shop  
3          Hockey Arena  
4           Coffee Shop

### And see how many venues were returned for each neighbourhood.

In [24]:
pd.set_option('display.max_columns', None)     # display all columns
pd.set_option('display.max_rows', None)        # display all rows

toronto_venues.groupby('Neighborhood').count()

Neighborhood Latitude  \
Neighborhood                                                                
Agincourt                                                               4   
Alderwood, Long Branch                                                  8   
Bathurst Manor, Wilson Heights, Downsview North                        23   
Bayview Village                                                         4   
Bedford Park, Lawrence Manor East                                      24   
Berczy Park                                                            56   
Birch Cliff, Cliffside West                                             4   
Brockton, Parkdale Village, Exhibition Place                           23   
CN Tower, King and Spadina, Railway Lands, Harb...                     16   
Caledonia-Fairbanks                                                     4   
Cedarbrae                                                               9   
Central Bay Street                                                     66   
Christie                                                               16   
Church and Wellesley                                                   80   
Clairville, Humberwood, Woodbine Downs, West Hu...                      4   
Clarks Corners, Tam O'Shanter, Sullivan                                13   
Cliffside, Cliffcrest, Scarborough Village West                         2   
Commerce Court, Victoria Hotel                                        100   
Davisville                                                             35   
Davisville North                                                        8   
Del Ray, Mount Dennis, Keelsdale and Silverthorn                        5   
Don Mills North                                                         4   
Don Mills South                                                        20   
Dorset Park, Wexford Heights, Scarborough Town ...                      5   
Downsview Central                                                       3   
Downsview East                                                          2   
Downsview Northwest                                                     4   
Downsview West                                                          6   
Dufferin, Dovercourt Village                                           16   
Enclave of L4W                                                         13   
Enclave of M4L                                                         16   
Enclave of M5E                                                         97   
Eringate, Bloordale Gardens, Old Burnhamthorpe,...                      8   
Fairview, Henry Farm, Oriole                                           65   
First Canadian Place, Underground city                                100   
Forest Hill North & West                                                4   
Garden District, Ryerson                                              100   
Glencairn                                                               5   
Golden Mile, Clairlea, Oakridge                                        10   
Guildwood, Morningside, West Hill                                       9   
Harbourfront East, Union Station, Toronto Islands                     100   
High Park, The Junction South                                          24   
Hillcrest Village                                                       6   
Humber Summit                                                           1   
Humberlea, Emery                                                        1   
Humewood-Cedarvale                                                      3   
India Bazaar, The Beaches West                                         21   
Kennedy Park, Ionview, East Birchmount Park                             6   
Kensington Market, Chinatown, Grange Park                              66   
Kingsview Village, St. Phillips, Martin Grove G...                      4   
Lawrence Manor, Lawrence Heights                

### Let's find out how many unique categories we have.

In [25]:
print('There are {} unique categories.'.format(len(toronto_venues['Venue Category'].unique())))

There are 271 unique categories.


### Now we analyze each neighbourhood.

In [26]:
# use one hot encoding to create dummy variables for each neighbourhood-category combination
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")

# then add the neighbourhood column back into the dataframe
toronto_onehot['Neighborhood'] = toronto_venues['Neighborhood']

# and move the neighbourhood column to be the first column in the dataframe
col_name = 'Neighborhood'
first_col = toronto_onehot.pop('Neighborhood')
toronto_onehot.insert(0, col_name, first_col)

# check the dataframe
toronto_onehot.head()


Neighborhood  Accessories Store  Adult Boutique  Airport  \
0         Parkwoods                  0               0        0   
1         Parkwoods                  0               0        0   
2         Parkwoods                  0               0        0   
3  Victoria Village                  0               0        0   
4  Victoria Village                  0               0        0   

   Airport Food Court  Airport Gate  Airport Lounge  Airport Service  \
0                   0             0               0                0   
1                   0             0               0                0   
2                   0             0               0                0   
3                   0             0               0                0   
4                   0             0               0                0   

   Airport Terminal  American Restaurant  Antique Shop  Aquarium  Art Gallery  \
0                 0                    0             0         0            0   
1                 0                    0             0         0            0   
2                 0                    0             0         0            0   
3                 0                    0             0         0            0   
4                 0                    0             0         0            0   

   Art Museum  Arts & Crafts Store  Asian Restaurant  Athletics & Sports  \
0           0                    0                 0                   0   
1           0                    0                 0                   0   
2           0                    0                 0                   0   
3           0                    0                 0                   0   
4           0                    0                 0                   0   

   Auto Garage  Auto Workshop  BBQ Joint  Baby Store  Bagel Shop  Bakery  \
0            0              0          0           0           0       0   
1            0              0          0           0           0       0   
2            0              0          0           0           0       0   
3            0              0          0           0           0       0   
4            0              0          0           0           0       0   

   Bank  Bar  Baseball Field  Baseball Stadium  Basketball Stadium  Beach  \
0     0    0               0                 0                   0      0   
1     0    0               0                 0                   0      0   
2     0    0               0                 0                   0      0   
3     0    0               0                 0                   0      0   
4     0    0               0                 0                   0      0   

   Bed & Breakfast  Beer Bar  Beer Store  Belgian Restaurant  Bike Shop  \
0                0         0           0                   0          0   
1                0         0           0                   0          0   
2                0         0           0                   0          0   
3                0         0           0                   0          0   
4                0         0           0                   0          0   

   Bistro  Boat or Ferry  Bookstore  Boutique  Brazilian Restaurant  \
0       0              0          0         0                     0   
1       0              0          0         0                     0   
2       0              0          0         0                     0   
3       0              0          0         0                     0   
4       0              0          0         0                     0   

   Breakfast Spot  Brewery  Bridal Shop  Bubble Tea Shop  Building  \
0               0        0            0                0         0   
1               0        0            0                0         0   
2               0        0            0                0         0   
3               0        0            0                0         0   
4               0        0            0                0         0   

   Burge

### Before we had 2141 rows and 7 columns. We should now have the same number of rows but many more columns now.

In [27]:
toronto_onehot.shape

(2141, 271)

### We're going to create clusters based on similarity in venue category. First we group group rows by neighbourhood and take the mean of the frequency of occurance of each category.

In [28]:
toronto_grouped = toronto_onehot.groupby('Neighborhood').mean().reset_index()
toronto_grouped

Neighborhood  Accessories Store  \
0                                           Agincourt           0.000000   
1                              Alderwood, Long Branch           0.000000   
2     Bathurst Manor, Wilson Heights, Downsview North           0.000000   
3                                     Bayview Village           0.000000   
4                   Bedford Park, Lawrence Manor East           0.000000   
5                                         Berczy Park           0.000000   
6                         Birch Cliff, Cliffside West           0.000000   
7        Brockton, Parkdale Village, Exhibition Place           0.000000   
8   CN Tower, King and Spadina, Railway Lands, Har...           0.000000   
9                                 Caledonia-Fairbanks           0.000000   
10                                          Cedarbrae           0.000000   
11                                 Central Bay Street           0.000000   
12                                           Christie           0.000000   
13                               Church and Wellesley           0.000000   
14  Clairville, Humberwood, Woodbine Downs, West H...           0.000000   
15            Clarks Corners, Tam O'Shanter, Sullivan           0.000000   
16    Cliffside, Cliffcrest, Scarborough Village West           0.000000   
17                     Commerce Court, Victoria Hotel           0.000000   
18                                         Davisville           0.000000   
19                                   Davisville North           0.000000   
20   Del Ray, Mount Dennis, Keelsdale and Silverthorn           0.000000   
21                                    Don Mills North           0.000000   
22                                    Don Mills South           0.000000   
23  Dorset Park, Wexford Heights, Scarborough Town...           0.000000   
24                                  Downsview Central           0.000000   
25                                     Downsview East           0.000000   
26                                Downsview Northwest           0.000000   
27                                     Downsview West           0.000000   
28                       Dufferin, Dovercourt Village           0.000000   
29                                     Enclave of L4W           0.000000   
30                                     Enclave of M4L           0.000000   
31                                     Enclave of M5E           0.000000   
32  Eringate, Bloordale Gardens, Old Burnhamthorpe...           0.000000   
33                       Fairview, Henry Farm, Oriole           0.000000   
34             First Canadian Place, Underground city           0.000000   
35                           Forest Hill North & West           0.000000   
36                           Garden District, Ryerson           0.000000   
37                                          Glencairn           0.000000   
38                    Golden Mile, Clairlea, Oakridge           0.000000   
39                  Guildwood, Morningside, West Hill           0.000000   
40  Harbourfront East, Union Station, Toronto Islands           0.000000   
41                      High Park, The Junction South           0.000000   
42                                  Hillcrest Village           0.000000   
43                                      Humber Summit           0.000000   
44                                   Humberlea, Emery           0.000000   
45                                 Humewood-Cedarvale           0.000000   
46                     India Bazaar, The Beaches West           0.000000   
47        Kennedy Park, Ionview, East Birchmount Park           0.000000   
48          Kensington Market, Chinatown, Grange Park           0.000000   
49  Kingsview Village, St. Phillips, Martin Grove ...           0.000000   
50                   Lawrence Manor, Lawrence Heights           0.071429   
51                                      Lawrence Park           0.000000   
52           

### Let's do a quick check that we have the same number of columns as before, but fewer rows.

In [29]:
toronto_grouped.shape

(100, 271)

### We create a function to sort venue categories in descending order based on mean of frequency for each neighbourhood.

In [36]:
num_top_venues = 10

def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

### And then create a dataframe that can display the top 10 venue categories for each neighbourhood.

In [37]:
indicators = ['st', 'nd', 'rd']

# create columns in line with num_top_venues
columns = ['Neighborhood']                                                                # start with neighbourhoods dataframe
for ind in np.arange(num_top_venues):                                                     # for each ind in the num_top_venues range
    try:                 
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))           # append ind value in first {} and indicators value in second {}
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))                            # except when there is no indicator value, then just put 'th' 

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)                               # dataframe where the columns are the ones created above
neighborhoods_venues_sorted['Neighborhood'] = toronto_grouped['Neighborhood']             # add a column 'Neighborhood' to the df - identical to the one we had

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()


Neighborhood 1st Most Common Venue  \
0                                        Agincourt                Lounge   
1                           Alderwood, Long Branch           Pizza Place   
2  Bathurst Manor, Wilson Heights, Downsview North           Coffee Shop   
3                                  Bayview Village                  Café   
4                Bedford Park, Lawrence Manor East        Sandwich Place   

  2nd Most Common Venue      3rd Most Common Venue    4th Most Common Venue  \
0        Breakfast Spot  Latin American Restaurant           Clothing Store   
1              Pharmacy                Coffee Shop                      Pub   
2                  Bank             Ice Cream Shop       Frozen Yogurt Shop   
3   Japanese Restaurant         Chinese Restaurant                     Bank   
4    Italian Restaurant                Coffee Shop  Comfort Food Restaurant   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0           Event Space    Falafel Restaurant  Ethiopian Restaurant   
1                  Pool        Sandwich Place                   Gym   
2        Sandwich Place                 Diner            Restaurant   
3                 Diner        Discount Store   Distribution Center   
4           Pizza Place             Juice Bar  Fast Food Restaurant   

  8th Most Common Venue 9th Most Common Venue     10th Most Common Venue  
0           Escape Room     Electronics Store                      Diner  
1   Dumpling Restaurant             Drugstore                 Donut Shop  
2          Intersection           Supermarket  Middle Eastern Restaurant  
3               Dog Run      Doner Restaurant                Yoga Studio  
4               Butcher                  Café                 Restaurant

### We cluster the neighbourhoods.

In [79]:
# set the number of clusters
kclusters = 3

# neighbourhood is not an int column so we drop it for running k-means
toronto_grouped_clustering = toronto_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10]


array([0, 0, 0, 0, 0, 0, 0, 0, 0, 1])

### And create a new dataframe that has all locations and the cluster label based on neighborhood.

In [81]:
# add cluster labels to neighborhoods_venues_sorted
#neighborhoods_venues_sorted = neighborhoods_venues_sorted.drop('Cluster Labels', 1)         # uncomment if changing the number of cluster
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

toronto_merged = df_with_coords                                                              # the original df

# add neighborhoods_venues_sorted (now with cluster labels) to the original df
toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')

toronto_merged.head()


PostalCode           Borough                      Neighborhood   Latitude  \
0        M3A        North York                         Parkwoods  43.753259   
1        M4A        North York                  Victoria Village  43.725882   
2        M5A  Downtown Toronto         Regent Park, Harbourfront  43.654260   
3        M6A        North York  Lawrence Manor, Lawrence Heights  43.718518   
4        M7A      Queen's Park     Ontario Provincial Government  43.662301   

   Longitude  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
0 -79.329656             1.0  Fast Food Restaurant     Food & Drink Shop   
1 -79.315572             0.0          Intersection           Pizza Place   
2 -79.360636             0.0           Coffee Shop                  Park   
3 -79.464763             0.0        Clothing Store     Accessories Store   
4 -79.389494             0.0           Coffee Shop      Sushi Restaurant   

  3rd Most Common Venue 4th Most Common Venue  5th Most Common Venue  \
0                  Park        Discount Store    Distribution Center   
1     French Restaurant           Coffee Shop  Portuguese Restaurant   
2                Bakery                   Pub         Breakfast Spot   
3           Coffee Shop   Sporting Goods Shop             Shoe Store   
4                  Bank                   Bar               Beer Bar   

    6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
0                 Dog Run      Doner Restaurant            Donut Shop   
1            Hockey Arena      Doner Restaurant    Dim Sum Restaurant   
2                    Café               Theater            Beer Store   
3  Furniture / Home Store              Boutique   Arts & Crafts Store   
4                     Spa         Smoothie Shop         Burrito Place   

  9th Most Common Venue 10th Most Common Venue  
0             Drugstore    Dumpling Restaurant  
1                 Diner         Discount Store  
2                   Spa             Shoe Store  
3           Event Space          Women's Store  
4        Sandwich Place                   Café

In [82]:
# check for missing values
toronto_merged.isnull().sum()

PostalCode                0
Borough                   0
Neighborhood              0
Latitude                  0
Longitude                 0
Cluster Labels            3
1st Most Common Venue     3
2nd Most Common Venue     3
3rd Most Common Venue     3
4th Most Common Venue     3
5th Most Common Venue     3
6th Most Common Venue     3
7th Most Common Venue     3
8th Most Common Venue     3
9th Most Common Venue     3
10th Most Common Venue    3
dtype: int64

In [83]:
toronto_merged.shape

(103, 16)

### Drop the three rows with missing values

In [84]:
toronto_merged = toronto_merged.dropna().reset_index(drop=True)          # note: redefine toronto_merged to reset the index (or not to not reset the index)
toronto_merged.shape


(100, 16)

### Make a map with different colours for each cluster.

In [85]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged['Latitude'], toronto_merged['Longitude'], toronto_merged['Neighborhood'], toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[int(cluster)-1],
        fill=True,
        fill_color=rainbow[int(cluster)-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

### Let's have a closer look at each cluster.

In [70]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 0, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
20              York             0.0                  Park   
39        North York             0.0               Airport   
50        North York             0.0                  Park   
62              York             0.0                  Park   
64        North York             0.0                  Park   
89  Downtown Toronto             0.0                  Park   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
20         Women's Store                  Pool             Drugstore   
39                  Park           Yoga Studio             Drugstore   
50           Yoga Studio             Drugstore                 Diner   
62           Yoga Studio             Drugstore                 Diner   
64     Convenience Store           Yoga Studio             Drugstore   
89            Playground                 Trail           Yoga Studio   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
20                 Diner        Discount Store   Distribution Center   
39                 Diner        Discount Store   Distribution Center   
50        Discount Store   Distribution Center               Dog Run   
62        Discount Store   Distribution Center               Dog Run   
64        Discount Store   Distribution Center               Dog Run   
89      Doner Restaurant    Dim Sum Restaurant                 Diner   

   8th Most Common Venue 9th Most Common Venue       10th Most Common Venue  
20               Dog Run      Doner Restaurant                   Donut Shop  
39               Dog Run      Doner Restaurant                   Donut Shop  
50      Doner Restaurant            Donut Shop          Dumpling Restaurant  
62      Doner Restaurant            Donut Shop          Dumpling Restaurant  
64      Doner Restaurant            Donut Shop  Eastern European Restaurant  
89        Discount Store   Distribution Center                      Dog Run

In [86]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 1, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
0               North York             1.0  Fast Food Restaurant   
20                    York             1.0                  Park   
34  East York/East Toronto             1.0          Intersection   
39              North York             1.0               Airport   
47              North York             1.0                  Park   
59         Central Toronto             1.0                  Park   
66         Central Toronto             1.0                  Park   
75               Etobicoke             1.0        Sandwich Place   
81         Central Toronto             1.0                   Gym   
83             Scarborough             1.0            Playground   
89        Downtown Toronto             1.0                  Park   
95               Etobicoke             1.0                  Park   

         2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0            Food & Drink Shop                  Park        Discount Store   
20               Women's Store                  Pool             Drugstore   
34                        Park     Convenience Store           Yoga Studio   
39                        Park           Yoga Studio             Drugstore   
47  Construction & Landscaping                Bakery        Massage Studio   
59            Business Service              Bus Line           Swim School   
66                       Trail      Sushi Restaurant         Jewelry Store   
75                        Park     Mobile Phone Shop              Bus Line   
81                        Park                 Trail           Yoga Studio   
83                        Park          Intersection           Yoga Studio   
89                  Playground                 Trail           Yoga Studio   
95                       River                  Pool           Yoga Studio   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0    Distribution Center               Dog Run      Doner Restaurant   
20                 Diner        Discount Store   Distribution Center   
34             Drugstore        Discount Store   Distribution Center   
39                 Diner        Discount Store   Distribution Center   
47           Yoga Studio             Drugstore   Distribution Center   
59            Donut Shop   Distribution Center               Dog Run   
66            Donut Shop        Discount Store   Distribution Center   
75      Doner Restaurant                 Diner        Discount Store   
81            Donut Shop                 Diner        Discount Store   
83            Donut Shop                 Diner        Discount Store   
89      Doner Restaurant    Dim Sum Restaurant                 Diner   
95      Doner Restaurant    Dim Sum Restaurant                 Diner   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0             Donut Shop             Drugstore    Dumpling Restaurant  
20               Dog Run      Doner Restaurant             Donut Shop  
34               Dog Run      Doner Restaurant             Donut Shop  
39               Dog Run      Doner Restaurant             Donut Shop  
47               Dog Run      Doner Restaurant             Donut Shop  
59      Doner Restaurant             Drugstore                  Diner  
66               Dog Run      Doner Restaurant              Drugstore  
75   Distribution Center               Dog Run            Yoga Studio  
81   Distribution Center               Dog Run       Doner Restaurant  
83   Distribution Center               Dog Run       Doner Restaurant  
89        Discount Store   Distribution Center                Dog Run  
95        Discount Store   Distribution Center                Dog Run

In [87]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 2, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
50  North York             2.0                  Park           Yoga Studio   
62        York             2.0                  Park           Yoga Studio   
64  North York             2.0                  Park     Convenience Store   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
50             Drugstore                 Diner        Discount Store   
62             Drugstore                 Diner        Discount Store   
64           Yoga Studio             Drugstore        Discount Store   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
50   Distribution Center               Dog Run      Doner Restaurant   
62   Distribution Center               Dog Run      Doner Restaurant   
64   Distribution Center               Dog Run      Doner Restaurant   

   9th Most Common Venue       10th Most Common Venue  
50            Donut Shop          Dumpling Restaurant  
62            Donut Shop          Dumpling Restaurant  
64            Donut Shop  Eastern European Restaurant

### Toronto appears to have a lot of parks, yoga studios and drug stores...